src/data/pi/lex.lean

```diff
@@ -7,33 +7,44 @@ import order.well_founded
 import algebra.group.pi
 import order.min_max
 
-
 /-!
 # Lexicographic order on Pi types
 
-This file defines the lexicographic relation for Pi types of partial orders and linear orders. We
-also provide a `pilex` analog of `pi.ordered_comm_group` (see `algebra.order.pi`).
+This file defines the lexicographic order for Pi types. `a` is less than `b` if `a i = b i` for all
+`i` up to some point `k`, and `a k < b k`.
+
+## Notation
+
+* `Πₗ i, α i`: Pi type equipped with the lexicographic order. Type synonym of `Π i, α i`.
+
+## See also
+
+Related files are:
+* `data.finset.colex`: Colexicographic order on finite sets.
+* `data.list.lex`: Lexicographic order on lists.
+* `data.sigma.order`: Lexicographic order on `Σₗ i, α i`.
+* `data.psigma.order`: Lexicographic order on `Σₗ' i, α i`.
+* `data.prod.lex`: Lexicographic order on `α × β`.
 -/
 
 variables {ι : Type*} {β : ι → Type*} (r : ι → ι → Prop)
   (s : Π {i}, β i → β i → Prop)
 
+namespace pi
+
+instance {α : Type*} : Π [inhabited α], inhabited (lex α) := id
+
 /-- The lexicographic relation on `Π i : ι, β i`, where `ι` is ordered by `r`,
   and each `β i` is ordered by `s`. -/
-def pi.lex (x y : Π i, β i) : Prop :=
+protected def lex (x y : Π i, β i) : Prop :=
 ∃ i, (∀ j, r j i → x j = y j) ∧ s (x i) (y i)
 
-/-- The cartesian product of an indexed family, equipped with the lexicographic order. -/
-def pilex (α : Type*) (β : α → Type*) : Type* := Π a, β a
+notation `Πₗ` binders `, ` r:(scoped p, lex (Π i, p i)) := r
```

Comment:
Should this be
```suggestion
notation `Πₗ ` binders `, ` r:(scoped p, lex (Π i, p i)) := r
```
or does lean add the space automaticaly?

Reply:
It seems like it does. This is the infoview on line 147 of `data.sigma.order`.
```
_x : Σₗ (i : ι), α i
```
and this is the notation declaration
```
notation `Σₗ` binders `, ` r:(scoped p, _root_.lex (sigma p)) := r
```
However my new notation gets torn in the infoview.
```
a b : lex (Π (i : ι), (λ (i : ι), β i) i)
```
Is there an equivalent of `sigma p` (as opposed to `Σ i, p i`) for `Π i, p i`?

Reply:
> Is there an equivalent of `sigma p` (as opposed to `Σ i, p i`) for `Π i, p i`?

No, because if there were it would need two copies of itself in its own type!

Reply:
Ah yeah of course 😕

Reply:
Maybe best to not actually use the notation in this file then, so that the instances and lemmas are fully delta-reduced.

Reply:
I think the notation could still be useful within a proof if you're constructing some big nested lex object; the delta reductions are probably mostly harmless, but keeping them out of the basic API lemmas is worth the trouble.